### PR TITLE
 London | 26-Java-June | Pooriya Ketabi | Sprint 3 | Prep-Exercises

### PR DESCRIPTION
**Pull request by PooriyaKTB**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

Briefly explain your PR.

## Questions

Ask any questions you have for your reviewer. You must remove this section if you have no questions.

### CODE REVIEW

File: src/main/java/StringUtilities.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.List;
+
+/**
+ * Utility methods for working with strings.
+ */
+public class StringUtilities {
+
+    /**
+     * Checks whether a string is a palindrome.
+     * The comparison is case-insensitive.
```
Comment (jwjlord) [inline src/main/java/StringUtilities.java:11]: How does your palindrome checker handle punctuation and whitespace?

File: src/test/java/TestStringUtilities.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    @BeforeEach
+    void initialisation() {
+        stringUtilities = new StringUtilities();
+    }
+
+    @Test
+    @DisplayName("Test for isPalindrome()")
+    void testIsPalindrome() {
+
+        assertAll(
```
Comment (jwjlord) [inline src/test/java/TestStringUtilities.java:28]: very nice!

Comment (jwjlord) [inline src/test/java/TestStringUtilities.java:28]: Have you tried it for a palindrome with an even number of letters? What about words containing non-ASCII characters, such as été or état? Is 😀😊😀 a palindrome? Is "A man, a plan, a canal: Panama" a palindrome?

File: src/test/java/TestStringUtilities.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    @Test
+    @DisplayName("Test for splitCommaSeparatedLine()")
+    void testSplitCommaSeparatedLine() {
+
+        assertAll(
+                () -> assertEquals(List.of("java", "spring", "Junit", "121"), stringUtilities.splitCommaSeparatedLine("java ,spring, Junit,121")),
+                () -> assertEquals(List.of("java", "spring", "Junit"), stringUtilities.splitCommaSeparatedLine("java,spring,Junit ")),
+                () -> assertEquals(List.of("java", "", "Junit"), stringUtilities.splitCommaSeparatedLine("java,,Junit")),
+                () -> assertEquals(List.of("java", "Junit"), stringUtilities.splitCommaSeparatedLine("java,Junit")),
+                () -> assertEquals(List.of("", "", "java"), stringUtilities.splitCommaSeparatedLine(",,java")),
```
Comment (jwjlord) [inline src/test/java/TestStringUtilities.java:62]: Why did you decide that the argument ",,java" yields List.of("", "", "java") but the argument ",," yields an empty list, and "java,," yields List.of("java")? Seems somewhat inconsistent at first glance

File: src/test/java/TestStringUtilities.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                () -> assertEquals(List.of(), stringUtilities.splitCommaSeparatedLine(",,")),
+                () -> assertEquals(List.of(""), stringUtilities.splitCommaSeparatedLine("")),
+                () -> assertThrows(IllegalArgumentException.class, () -> stringUtilities.splitCommaSeparatedLine(null))
+        );
+    }
+
+
+//  *** Stretch Goal - Parameterised tests ***
+
+//  ** @ParameterizedTest with @ValueSource **
```
Comment (jwjlord) [inline src/test/java/TestStringUtilities.java:77]: Better way to annotate methods with comments?

File: src/test/java/TestCalculator.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    @BeforeEach
+    void setUp() {
+        calc = new Calculator();
+    }
+
+    @Test
+    @DisplayName("Test for add()")
+    void testAdd() {
+        assertAll(
+                () -> assertEquals(5, calc.add(2, 3)),
```
Comment (jwjlord) [inline src/test/java/TestCalculator.java:21]: Do you think it's worth testing what happens with negative arguments?

File: src/test/java/TestCalculator.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                () -> assertEquals(0, calc.add(0, 0)),
+                () -> assertEquals(1, calc.add(1, 0))
+        );
+    }
+
+    @Test
+    @DisplayName("Test for divide()")
+    void testDivide() {
+        assertAll(
+                () -> assertEquals(20, calc.divide(60, 3)),
```
Comment (jwjlord) [inline src/test/java/TestCalculator.java:34]: What happens when the divisor is not a factor of the dividend (for example, 7 divided by 3)?

File: src/test/java/TestCalculator.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                () -> assertEquals(10, calc.divide(100, 10)),
```
Comment (jwjlord) [inline src/test/java/TestCalculator.java:35]: What happens when the first argument is zero?

File: src/test/java/TestCalculator.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+                () -> assertEquals(0, calc.multiply(0, 25)),
+                () -> assertEquals(-25, calc.multiply(-1, 25))
+        );
+    }
+
+    @Test
+    @DisplayName("Test for subtract()")
+    void testSubtract() {
+        assertAll(
+                () -> assertEquals(1, calc.subtract(100, 99)),
```
Comment (jwjlord) [inline src/test/java/TestCalculator.java:68]: nice

File: src/test/java/TestCalculator.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+    /**
+     * Verifies square root calculations, including
+     * floating-point precision and invalid negative input.
+     */
+    @Test
+    @DisplayName("Test for squareRoot()")
+    void testSquareRoot() {
+        assertAll(
+                () -> assertEquals(3.0, calc.squareRoot(9)),
```
Comment (jwjlord) [inline src/test/java/TestCalculator.java:96]: nice

File: src/main/java/StringUtilities.java (hunk 1 above)
Comment (PooriyaKTB) [reply src/main/java/StringUtilities.java:11]: Good catch! As the exercise requirements didn't specify phrase-level handling, it currently performs an exact character-by-character check (case-insensitive only). This means whitespace and punctuation are treated as regular characters (like 'lev el' evaluates to false). If we want to support full phrases, we can filter out non-alphanumeric characters using regex or `Character.isLetterOrDigit()` before checking.

File: src/test/java/TestCalculator.java (hunk 5 above)
Comment (PooriyaKTB) [reply src/test/java/TestCalculator.java:21]: Good point. Adding negative cases like assertEquals(-5, calc.add(-2, -3)) and assertEquals(-3, calc.add(-5, 2)) helps 
I'll add test cases for negative arguments as well.
Thanks.

File: src/test/java/TestCalculator.java (hunk 6 above)
Comment (PooriyaKTB) [reply src/test/java/TestCalculator.java:34]: Since divide returns an int, Java drops the decimal part and discards the remainder (calc.divide(7, 3) returns 2).
I'll add test cases to cover non-factor divisions explicitly.

File: src/test/java/TestCalculator.java (hunk 7 above)
Comment (PooriyaKTB) [reply src/test/java/TestCalculator.java:35]: Good point. Dividing zero by any non-zero number returns 0 without throwing an exception.
I'll add `assertEquals(0, calc.divide(0, 5))` to the test suite to cover it.
Thank you.

File: src/test/java/TestCalculator.java (hunk 8 above)
Comment (PooriyaKTB) [reply src/test/java/TestCalculator.java:68]: Thanks.

File: src/test/java/TestCalculator.java (hunk 9 above)
Comment (PooriyaKTB) [reply src/test/java/TestCalculator.java:96]: Thank you.

File: src/test/java/TestStringUtilities.java (hunk 2 above)
Comment (PooriyaKTB) [reply src/test/java/TestStringUtilities.java:28]: Thanks.

Comment (PooriyaKTB) [reply src/test/java/TestStringUtilities.java:28]: Yes, exactly!
- "noon" and "été" evaluate to true.
- The emoji sequence evaluates to false because Java emojis consist of multiple char values, which break standard character comparison.
- The phrase "A man, a plan, a canal: Panama" evaluates to false because spaces and punctuation are treated as regular characters in the current logic.
I'll add tests for them!
Thanks for pointing them out.

File: src/test/java/TestStringUtilities.java (hunk 3 above)
Comment (PooriyaKTB) [reply src/test/java/TestStringUtilities.java:62]: This is due to Java's default `String.split(",")` which keeps leading/middle empty strings but discards trailing ones. We can pass a negative limit like `s.split(",", -1)` if we want to retain trailing empty values consistently.

File: src/test/java/TestStringUtilities.java (hunk 4 above)
Comment (PooriyaKTB) [reply src/test/java/TestStringUtilities.java:77]: Using JUnit's `@DisplayName` annotation or Javadoc would be much better for documenting the tests clearly than plain inline comments.
Thanks for mentioning it.
